service_configuration_lib/spark_config.py

```diff
@@ -162,66 +416,248 @@ def get_k8s_spark_env(
+
+    # configure sql shuffle partitions
+    spark_conf = _append_sql_shuffle_partitions_conf(spark_conf)
+    return spark_conf
+
+
+def parse_memory_string(memory_string: str) -> int:
+    return (
+        int(memory_string[:-1]) * 1024 if memory_string.endswith('g')
+        else int(memory_string)
+    )
+
+
+def get_signalfx_url(spark_conf: Mapping[str, str]) -> str:
+    return (
+        'https://app.signalfx.com/#/dashboard/DJzYJDkAcAA?density=4'
+        '&variables%5B%5D=Instance%3Dinstance_name:'
+        '&variables%5B%5D=Service%3Dservice_name:%5B%22spark%22%5D'
+        f"&variables%5B%5D=PaaSTA%20Cluster%3Dpaasta_cluster:{spark_conf['spark.executorEnv.PAASTA_CLUSTER']}"
+        f"&variables%5B%5D=PaaSTA%20Service%3Dpaasta_service:{spark_conf['spark.executorEnv.PAASTA_SERVICE']}"
+        f"&variables%5B%5D=PaaSTA%20Instance%3Dpaasta_instance:{spark_conf['spark.executorEnv.PAASTA_INSTANCE']}"
+        '&startTime=-6h&endTime=Now'
+    )
+
+
+def get_history_url(spark_conf: Mapping[str, str]) -> Optional[str]:
+    if spark_conf.get('spark.eventLog.enabled') != 'true':
+        return None
+    event_log_dir = spark_conf.get('spark.eventLog.dir')
+    with open(DEFAULT_SPARK_RUN_CONFIG) as fp:
+        spark_run_conf = yaml.safe_load(fp.read())
+    for env, env_conf in spark_run_conf['environments'].items():
+        if event_log_dir == env_conf['default_event_log_dir']:
+            return env_conf['history_server']
+    return None
+
+
+def get_resources_requested(spark_opts: Mapping[str, str]) -> Mapping[str, int]:
+    num_executors = (
+        # spark on k8s directly configure num instances
+        int(spark_opts.get('spark.executor.instances', 0)) or
+        # spark on mesos use cores.max and executor.core to calculate number of
+        # executors.
+        int(spark_opts.get('spark.cores.max', 0)) // int(spark_opts.get('spark.executor.cores', 0))
+    )
+    num_cpus = (
+        # spark on k8s
+        int(spark_opts.get('spark.executor.instances', 0)) * int(spark_opts.get('spark.executor.cores', 0)) or
+        # spark on mesos
+        int(spark_opts.get('spark.cores.max', 0))
+    )
+    num_gpus = int(spark_opts.get('spark.mesos.gpus.max', 0))
+
+    executor_memory = parse_memory_string(spark_opts.get('spark.executor.memory', ''))
+    # by default, spark adds an overhead of 10% of the executor memory, with a
+    # minimum of 384mb
+    memory_overhead: int = (
+        parse_memory_string(spark_opts['spark.executor.memoryOverhead'])
+        if spark_opts.get('spark.executor.memoryOverhead')
+        else max(384, int(0.1 * executor_memory))
+    )
+    total_memory = (executor_memory + memory_overhead) * num_executors
+
+    return {
+        'cpus': num_cpus,
+        'mem': total_memory,
+        # rough guess since spark does not collect this information
+        'disk': total_memory,
+        'gpus': num_gpus,
+    }
+
+
+def generate_clusterman_metrics_entries(
+    clusterman_metrics,
```

Comment:
Passing `clusterman_metrics` (i.e. a module) as an argument to where it is used is an unconventional way of having `service_configuration_lib` be able to use `clusterman_metrics` without needing to import it. When I suggested using an interface, I was envisioning more of a interface class that users (like `PaaSTA` and others) could initialize using `clusterman_metrics`. This interface pattern would be more in-line with being open-source.

That being said I think what you have here is okay, but I would include at minimum detailed comments explaining how this works and what `clusterman_metrics` is in this context, considering there is no import anywhere. I would also specify this is a Yelpism, as we make it pretty clear there are a number in this repo in the repo README.

Reply:
Thanks, I added docstring and code example for how to init the clusterman_metrics before passing it.  Please let me know if there's anything you'd like to add.